File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/_687/LongestUnivaluePath_687.java
package DataStructureAlgo.Java.LeetCode2025.ProblemSet.Trees._687;

import DataStructureAlgo.Java.helpers.TreeBuilder;
import DataStructureAlgo.Java.helpers.templates.TreeNode;

import java.util.Arrays;

/**
 * Author: Nitin Gupta
 * Date: 8/23/2024
 * Question Category:687. Longest Univalue Path
 * Description: https://leetcode.com/problems/longest-univalue-path/description
 * Given the root of a binary tree, return the length of the longest path, where each node in the path has the same value. This path may or may not pass through the root.
 *
 * The length of the path between two nodes is represented by the number of edges between them.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: root = [5,4,5,1,1,null,5]
 * Output: 2
 * Explanation: The shown image shows that the longest path of the same value (i.e. 5).
 * Example 2:
 *
 *
 * Input: root = [1,4,5,4,4,null,5]
 * Output: 2
 * Explanation: The shown image shows that the longest path of the same value (i.e. 4).
 *
 *
 * Constraints:
 *
 * The number of nodes in the tree is in the range [0, 104].
 * -1000 <= Node.val <= 1000
 * The depth of the tree will not exceed 1000.
 * File reference
 * -----------
 * Duplicate {@link DataStructureAlgo.Java.LeetCode.tree.LongestUniValuePath}
 * Similar {@link}
 * extension {@link }
 * <p>
 * Tags
 * -----
 * @medium
 * <p>
 * <p>
 * Company Tags
 * -----
 * @Amazon
 * @Apple
 * @Google
 *
 * @Editorial
 */
public class LongestUnivaluePath_687 {

    public static void main(String[] args) {
        boolean test = true;

        test &= test(new Integer[]{5,4,5,1,1,null,5}, 2);
        test &= test(new Integer[]{1,4,5,4,4,null,5}, 2);
        test &= test(new Integer[]{1,null,1,1,1,1,1,1}, 4);


        System.out.println("==========================");
        System.out.println("\nTest result = " + (test ? "Pass" : "Fail"));
    }

    private static boolean test(Integer[] input, int expected) {
        System.out.println("------------------------------");
        System.out.println("Input : "+ Arrays.toString(input) + "\nexpected : "+ expected);


        Solution solution = new Solution();
        int output = solution.longestUnivaluePath(TreeBuilder.buildTreeFromLevelOrder(input));

        boolean testResult = output == expected;
        System.out.println("output : "+ output + " testResult "+(testResult ? "Pass" : "Fail"));

        return testResult;
    }

    /**
     * We need to count the edges between the univalue path. THere is no edge to self-node, hence the unique path of each self-node would be 0;
     * We can start doing a post-order traversal,
     * 1. and if this is a leaf node, then its unique path length would be 0
     * 2. if this is not a leaf node, then we can compare its value to either left child or right child or both; for each equal value, it gains 1 length path.
     * 3. return maximum of either left or right to parent, as this will be a maximum edge we can add.
     * 4. Keep track of the maximum
     */
    static class Solution {
        public int longestUnivaluePath(TreeNode root) {
            if(root == null) return 0;

            int []max = {0};
            longestUnivaluePath(root, max);
            return max[0];

        }

        private int longestUnivaluePath(TreeNode root, int[]max){
            if(root == null || (root.right == null && root.left == null)) return 0;

            int leftUniValuePathLength = longestUnivaluePath(root.left, max);
            int rightUniValuePathLength = longestUnivaluePath(root.right, max);

            int leftPathLength = 0, rightPathLength = 0;
            if(root.left!=null && root.val == root.left.val)
                leftPathLength += leftUniValuePathLength + 1;
            if(root.right!=null && root.val == root.right.val)
                rightPathLength += rightUniValuePathLength + 1;

            max[0] = Math.max(max[0], leftPathLength + rightPathLength);
            return Math.max(leftPathLength, rightPathLength);
        }
    }
}

File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/_687/LongestUnivaluePath_687.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        """
        # COAST FRAMEWORK
        #
        # 1. CLARIFY:
        #    - We need longest path of SAME values.
        #    - Path length = number of EDGES.
        #    - Path can turn (inverted V shape) through a node.
        #
        # 2. OUTLINE:
        #    - Post-order DFS (bottom-up).
        #    - At each node, ask children: "What's the longest univalue path ending at you?"
        #    - If child.val == node.val, extend that path by 1 (the edge to child). root -> left / root -> right
        #    - If not, the path stops (length 0).
        #    - Update global max with (left_path + right_path) -> The "Bridge".
        #    - Return max(left_path, right_path) -> The "Arrow" to parent.
        #
        # 3. APPROACH (Edge-Based DFS):
        #    - Direct edge counting avoids 'ans - 1' hacks.
        """
        
        _max_length = 0

        def dfs(root):
            nonlocal _max_length
            if not root:
                return 0

            longest_left_length = dfs(root.left)
            longest_right_length = dfs(root.right)

            left_path_length_at_root, right_path_length_at_root = 0, 0 # path stops, if root.val != child.val
            
            
            if root.left and root.val == root.left.val:
                left_path_length_at_root = longest_left_length + 1  # one edge connected, root -> left

            if root.right and root.val == root.right.val:
                right_path_length_at_root = longest_right_length + 1 # one edge connected, root -> right

            _max_length = max(
                _max_length, left_path_length_at_root + right_path_length_at_root
            )

            return max(left_path_length_at_root, right_path_length_at_root)

        dfs(root)
        return _max_length

# COMPLEXITY ANALYSIS:
# Time: O(N) - Every node visited once.
# Space: O(H) - Recursion stack height.